sdk-extensions/jfr-events/src/main/java/io/opentelemetry/sdk/extension/jfr/JfrSpanProcessor.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.extension.jfr;
+
+import static java.util.Objects.nonNull;
+
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.sdk.common.CompletableResultCode;
+import io.opentelemetry.sdk.trace.ReadWriteSpan;
+import io.opentelemetry.sdk.trace.ReadableSpan;
+import io.opentelemetry.sdk.trace.SpanProcessor;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Span processor to create new JFR events for the Span as they are started, and commit on end.
+ *
+ * <p>NOTE: JfrSpanProcessor must be running synchronously to ensure that duration is correctly
+ * captured.
+ */
+public class JfrSpanProcessor implements SpanProcessor {
+
+  private final Map<SpanContext, SpanEvent> spanEvents = new ConcurrentHashMap<>();
+
+  @Override
+  public void onStart(Context parentContext, ReadWriteSpan span) {
+    if (span.getSpanContext().isValid()) {
+      SpanEvent event = new SpanEvent(span.toSpanData());
+      event.begin();
+      spanEvents.put(span.getSpanContext(), event);
+    }
+  }
+
+  @Override
+  public boolean isStartRequired() {
+    return true;
+  }
+
+  @Override
+  public void onEnd(ReadableSpan rs) {
+    SpanEvent event = spanEvents.remove(rs.getSpanContext());
+    if (nonNull(event) && event.shouldCommit()) {
+      event.commit();
+    }
+  }
+
+  @Override
+  public boolean isEndRequired() {
+    return true;
+  }
+
+  @Override
+  public CompletableResultCode shutdown() {
+    spanEvents.forEach((id, event) -> event.commit());
```

Comment:
Hmm - not really for this PR but this doesn't seem like how shutdown is supposed to work. For example, we would probably stop accepting spans, and wait for spans to end naturally in an exporter when it is shutting down I think. But can't think of any improvement here so just writing for reference.

Reply:
Probably would be good to improve the JavaDoc, not sure what 'not yet processed means'.

>    * Processes all span events that have not yet been processed and closes used resources.

Similar for forceFlush

>    * Processes all span events that have not yet been processed.

Sound more like this would be for finished spans rather than span that hasn't been closed yet.

Force flush would commit any ready events, as would shutdown but that would also clear up all resources, which I would take as drop all non-completed spans. Once something is shutdown I wouldn't expect things to continue to flow, unless the CompleteResultCode would wait for all spans, but that would potentially be indefinite.

will remove the event.commit part as that feels wrong here

Reply:
Agree we may need to improve docs, or otherwise expectations (maybe spec) for this (again not related to this PR). I would generally call this at the beginning of a graceful shutdown and expect pending requests, and exported spans, to complete before termination. /cc @jkwatson @oberon00

Reply:
what does "pending requests" mean here? Un-ended spans? Or just spans that were in the process of being sent off to exporters?